pkg/controller/storagecluster/initialization_reconciler.go

```diff
@@ -0,0 +1,374 @@
+package storagecluster
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	ocsv1 "github.com/openshift/ocs-operator/pkg/apis/ocs/v1"
+	"github.com/openshift/ocs-operator/pkg/controller/defaults"
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	corev1 "k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// ensureStorageClasses ensures that StorageClass resources exist in the desired
+// state.
+func (r *ReconcileStorageCluster) ensureStorageClasses(instance *ocsv1.StorageCluster, reqLogger logr.Logger) error {
+
+	if instance.Status.StorageClassesCreated {
+		return nil
+	}
+
+	scs, err := r.newStorageClasses(instance)
+	if err != nil {
+		return err
+	}
+	for _, sc := range scs {
+		err := controllerutil.SetControllerReference(instance, &sc, r.scheme)
+		if err != nil {
+			return err
+		}
+		existing := storagev1.StorageClass{}
+		err = r.client.Get(context.TODO(), types.NamespacedName{Name: sc.Name, Namespace: sc.Namespace}, &existing)
+
+		switch {
+		case err == nil:
+			reqLogger.Info(fmt.Sprintf("Restoring original StorageClass %s", sc.Name))
+			sc.ObjectMeta = existing.ObjectMeta
```

Comment:
What if existing StorageClass is marked for deletion?
If we want to restore it then wouldn't we need to remove DeletionTimestamp?
Cache could cause issues.

Reply:
we can't remove the DeletionTimestamp.

you're right we need to account for objects marked for deletion though. great catch :+1:  I'll make this change.